acbuild/script_test.go

```diff
@@ -0,0 +1,251 @@
+		output := joinLines(c.input)
+		if !equal(output, c.output) {
+			t.Errorf("output, expected:%v actual:%v", c.output, output)
+		}
+	}
+}
+
+func TestTokenizeLine(t *testing.T) {
+	type testcase struct {
+		input  string
+		output []string
+		err    error
+	}
+	cases := []testcase{
+		// nothing special here
+		testcase{
+			"run -- test",
+			[]string{"run", "--", "test"},
+			nil,
+		},
+
+		// whitespace shenanigans
+		testcase{
+			"run --  test",
+			[]string{"run", "--", "test"},
+			nil,
+		},
+		testcase{
+			"run -- 	 test",
+			[]string{"run", "--", "test"},
+			nil,
+		},
+
+		// can it escape slashes?
+		testcase{
+			`set-name \\'hello'\\""`,
+			[]string{"set-name", `\hello\`},
+			nil,
+		},
+
+		// do single quotes behave as expected?
+		testcase{
+			`set-name \'hello\'`,
+			[]string{"set-name", "'hello'"},
+			nil,
+		},
+		testcase{
+			`set-name 'hello'`,
+			[]string{"set-name", "hello"},
+			nil,
+		},
+		testcase{
+			`set-name 'he llo'`,
+			[]string{"set-name", "he llo"},
+			nil,
+		},
+		testcase{
+			`set-name \'he llo\'`,
+			[]string{"set-name", "'he", "llo'"},
+			nil,
+		},
+
+		// do double quotes behave as expected?
+		testcase{
+			`set-name \"hello\"`,
+			[]string{"set-name", `"hello"`},
+			nil,
+		},
+		testcase{
+			`set-name "hello"`,
+			[]string{"set-name", "hello"},
+			nil,
+		},
+		testcase{
+			`set-name "he llo"`,
+			[]string{"set-name", "he llo"},
+			nil,
+		},
+		testcase{
+			`set-name \"he llo\"`,
+			[]string{"set-name", `"he`, `llo"`},
+			nil,
+		},
+
+		// can I properly nest quotes?
+		testcase{
+			`set-name "t'es't"`,
+			[]string{"set-name", "t'es't"},
+			nil,
+		},
+		testcase{
+			`set-name "t'est"`,
+			[]string{"set-name", "t'est"},
+			nil,
+		},
+		testcase{
+			`set-name 't"es"t'`,
+			[]string{"set-name", `t"es"t`},
+			nil,
+		},
+		testcase{
+			`set-name 'tes"t'`,
+			[]string{"set-name", `tes"t`},
+			nil,
+		},
+
+		// will it error when it should?
+		testcase{
+			`set-name 'test"`,
+			[]string{},
+			errSingleQuote,
+		},
+		testcase{
+			`set-name "test'`,
+			[]string{},
+			errDoubleQuote,
+		},
+		testcase{
+			`set-name test\`,
+			[]string{},
+			errEscape,
+		},
+
+		// are quotes in the middle of words handled?
+		testcase{
+			`set-name t"es"t`,
+			[]string{"set-name", "test"},
+			nil,
+		},
+		testcase{
+			`set-name t'es't`,
+			[]string{"set-name", "test"},
+			nil,
+		},
+		testcase{
+			`set-name "te"'st'`,
+			[]string{"set-name", "test"},
+			nil,
+		},
+
+		// let's try something crazy
+		testcase{
+			`set-name 't'h"i's "i\'s" 'a' "t''''e""""s'""'t\\`,
+			[]string{"set-name", `thi's i's 'a' tes""t\`},
+			nil,
+		},
+	}
+
+	for _, c := range cases {
+		output, err := tokenizeLine(c.input)
+		if err != c.err {
+			t.Errorf("error, expected:%v actual:%v", c.err, err)
+		}
+		if !equal(output, c.output) {
+			t.Errorf("output, expected:%v actual:%v", c.output, output)
+		}
+	}
+}
+
+// no really guys, this language is _great_
```

Comment:
People always ask me, what about languages? I very quicky answer, we have the best languages, they're outstanding.